firmware.py

```diff
@@ -63,11 +69,63 @@ def details(company_id):
+    category = categoryCursors.fetchall()
     reviewsCursor = db.execute(
         'select user, review from reviews where company_id="%s" order by id desc'
         % company_id
     )
     reviews = reviewsCursor.fetchall()
-    return render_template('details.html', reviews=reviews, company=records[0])
+    return render_template('details.html', reviews=reviews, company=records[0], category=category[0]['category_type'])
+
+
+@app.route('/add/company', methods=['GET', 'POST'])
+def add_company():
+    db = get_db()
+    submitted_data = request;
+    if request.method == 'POST':
+        submitted_data = request
+        a = request.form['company_name']
+        if(a == ""):
+            flash("Please add a name for your company!")
+            return render_template('add_company.html', data=submitted_data)
+        e = request.form['company_adress']
+        if(e == ""):
+            flash("Please add an adress for your company!")
+            return render_template('add_company.html', data=submitted_data)
+        b = request.form['company_description']
+        if(b == ""):
+            flash("Please add a short description for your company!")
+            return render_template('add_company.html', data=submitted_data)
+        if request.form.get('select-category-list') == None:
+            flash('Please select a category for your company!')
+            return render_template('add_company.html', data=submitted_data)
+        addCursor = db.execute("select category_id from category where \
+            category_type='%s'" % str(request.form.get('select-category-list')))
+        cat_id = addCursor.fetchall()[0]
+        c = request.form['company_details']
+        if(c == ""):
+            flash("Please add a few details about your company!")
+            return render_template('add_company.html', data=submitted_data)
+        d = str(int(cat_id[0]))
+        file = request.files['company_logo']
+        if file:
+            filename = file.filename
+            file.save(os.path.join(app.config['UPLOAD_FOLDER'],filename))
+            os.rename(UPLOAD_FOLDER+filename, UPLOAD_FOLDER+a+".jpg")
+        else:
+            flash("No logo added for your company.You can do that later!")
+            filename = "default"
+        db.execute("insert into companies (company_name, \
+            company_description , company_details, company_rating, \
+            company_logo, company_adress, category_id) values \
+            (?, ?, ?, ?, ?, ?, ?)", (a, b, c, 0, filename+".jpg", e, d)
+            )
+        db.commit()
+        flash('Congratulations on adding your company, '+a+' to our website!Check out your profile below.')
+        addCursor = db.execute("select MAX(company_id) from companies")
+        cat_id = addCursor.fetchall()[0]
+        return redirect(url_for('details', company_id=cat_id[0]))
+    return render_template('add_company.html', data=request)
```

Comment:
Do not send to the view the whole request object. Send only what you need, which is the form object `request.form`